### ISSUE TITLE
Add Title of File to the OLP

### ISSUE THREAD

**Issue by publicimageltd** [org-roam/org-roam#1956]
### Brief Abstract

I would like to know where a node "belongs to" if it is a heading. The data base field OLP seems to contain an outline path, but it begins at level 1 headings. I propose to include the title of the containing file as a level 0 heading, too. This way, an outline path can be easily displayed.

### Proposed Implementation (if any)

```elisp
(cl-defun org-roam-db-insert-node-data ()
....
(olp (append
      (list (org-link-display-format
              (or (cadr (assoc "TITLE" (org-collect-keywords '("title"))
                               #'string-equal))
                  (file-name-sans-extension
                   (file-relative-name file org-roam-directory)))))
      (org-get-outline-path nil 'use-cache)))
....
```

### Please check the following:

- [x] No similar feature requests

**Comment by jethrokuan** [comment]
Please see this wiki entry: https://github.com/org-roam/org-roam/wiki/Hitchhiker's-Rough-Guide-to-Org-roam-V2#showing-node-hierarchy

**Comment by publicimageltd** [comment]
Thanks for the link to the wiki. That's a fine approach, but it is too costly for longer queries since it requires to handle each node separately. `org-roam-filetitle` also requires the buffer to be loaded, if I understand it correctly, since `org-roam-get-keyword` presupposes to be in the active org mode buffer. So I'd need to loop through all nodes and even files to additionally add the filetitle to each node, right? My use case, however, is displaying nodes by querying them directly from the database (see [here](https://github.com/publicimageltd/delve)), and I would rather want to avoid looping over all nodes once more.

To put it differently: Isn't the method in the wiki relatively slow, so that it might be generally of benefit to add the file title to the OLP from begin with? I would do PR once you agree that it's worth it.

**Comment by jethrokuan** [comment]
You're right that the wiki method is costly.

I'm disagreeable to modifying directly what the database stores as OLP: we should keep this as authentic to Org as possible. The file title is not part of the OLP.

I am agreeable to this approach instead, and open to a PR:

1. Add a column in the `files` table for `file-title`
2. Additionally caching the file-title in `org-roam-db-insert-file`
3. Adding an additional property `file-title` to the `nodes` structure
4. Ensure that the `file-title` property is properly populated in `org-roam-populate` and `org-roam-node-list`

Then in your own user config you can create your own custom `org-roam-node-olp-with-title` function that's just a string concatenation of the two and use it in your `org-roam-node-display-template`.

**Comment by publicimageltd** [comment]
OK, PR will be coming. Looking at the code, two questions:

1. On `org-roam-db.el`, sometimes the title of the file is wrapped in `org-link-display-format`. Why?
2. In `org-roam-db-insert-file-node`, if there is no title, a provisory title is returned created from the file name. I understood, however, that files without a title are not allowed anymore. Shouldn' t I just remove that fallback?

**Comment by jethrokuan** [comment]
1. This is to convert the string title into a human readable format if it contains links in them (common in headings)
2. No, title is still optional, just highly encouraged.

**Comment by publicimageltd** [comment]
Thanks! Never thought of title containing links :smile:

**Comment by jethrokuan** [comment]
Closed with #1963
